packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts

```diff
@@ -1,20 +1,340 @@
 import { BigNumber } from '@0x/utils';
 import * as _ from 'lodash';
+import { performance } from 'perf_hooks';
 
-import { MarketOperation } from '../../types';
+import { DEFAULT_INFO_LOGGER } from '../../constants';
+import { MarketOperation, NativeOrderWithFillableAmounts } from '../../types';
+import { VIP_ERC20_BRIDGE_SOURCES_BY_CHAIN_ID, ZERO_AMOUNT } from '../market_operation_utils/constants';
 
+import { dexSamplesToFills, nativeOrdersToFills } from './fills';
 import { DEFAULT_PATH_PENALTY_OPTS, Path, PathPenaltyOpts } from './path';
-import { ERC20BridgeSource, Fill } from './types';
+import { getRate } from './rate_utils';
+import { DexSample, ERC20BridgeSource, FeeSchedule, Fill, FillData } from './types';
 
 // tslint:disable: prefer-for-of custom-no-magic-numbers completed-docs no-bitwise
 
 const RUN_LIMIT_DECAY_FACTOR = 0.5;
+const RUST_ROUTER_NUM_SAMPLES = 200;
+const FILL_QUOTE_TRANSFORMER_GAS_OVERHEAD = new BigNumber(150e3);
+// NOTE: The Rust router will panic with less than 3 samples
+const MIN_NUM_SAMPLE_INPUTS = 3;
+
+const isDexSample = (obj: DexSample | NativeOrderWithFillableAmounts): obj is DexSample => !!(obj as DexSample).source;
+
+function nativeOrderToNormalizedAmounts(
+    side: MarketOperation,
+    nativeOrder: NativeOrderWithFillableAmounts,
+): { input: BigNumber; output: BigNumber } {
+    const { fillableTakerAmount, fillableTakerFeeAmount, fillableMakerAmount } = nativeOrder;
+    const makerAmount = fillableMakerAmount;
+    const takerAmount = fillableTakerAmount.plus(fillableTakerFeeAmount);
+    const input = side === MarketOperation.Sell ? takerAmount : makerAmount;
+    const output = side === MarketOperation.Sell ? makerAmount : takerAmount;
+
+    return { input, output };
+}
+
+function calculateOuputFee(
+    side: MarketOperation,
+    sampleOrNativeOrder: DexSample | NativeOrderWithFillableAmounts,
+    outputAmountPerEth: BigNumber,
+    inputAmountPerEth: BigNumber,
+    fees: FeeSchedule,
+): BigNumber {
+    if (isDexSample(sampleOrNativeOrder)) {
+        const { source, fillData } = sampleOrNativeOrder;
+        const fee = fees[source] === undefined ? 0 : fees[source]!(fillData) || 0;
+        const outputFee = !outputAmountPerEth.isZero()
+            ? outputAmountPerEth.times(fee)
+            : inputAmountPerEth
+                  .times(fee)
+                  .times(sampleOrNativeOrder.output.dividedToIntegerBy(sampleOrNativeOrder.input));
+
+        return outputFee;
+    } else {
+        const { input, output } = nativeOrderToNormalizedAmounts(side, sampleOrNativeOrder);
+        const fee =
+            fees[ERC20BridgeSource.Native] === undefined ? 0 : fees[ERC20BridgeSource.Native]!(sampleOrNativeOrder);
+        const outputFee = !outputAmountPerEth.isZero()
+            ? outputAmountPerEth.times(fee)
+            : inputAmountPerEth.times(fee).times(output.dividedToIntegerBy(input));
+        return outputFee;
+    }
+}
+
+function createPathFromSamples(
```

Comment:
This function feels like it has a deceptively simple name considering all it does.

Reply:
yeah that's fine